pkg/io/option.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package io
+
+import (
+	"cloud.google.com/go/storage"
+	"gocloud.dev/blob"
+	"google.golang.org/api/googleapi"
+	utilpointer "k8s.io/utils/pointer"
+)
+
+// WriterOptions are options for the opener Writer method
+type WriterOptions struct {
+	BufferSize      *int64
+	ContentEncoding *string
+	ContentType     *string
+	Metadata        *map[string]string
+}
+
+// Apply applies the WriterOptions to storage.Writer and blob.WriterOptions
+// Both arguments are allowed to be nil.
+func (wo WriterOptions) Apply(writer *storage.Writer, o *blob.WriterOptions) {
```

Comment:
Oh, probably because you want to do this?

```golang

type WriterOptions struct {
  BufferSize int64
}

// NOTE: receiver should be a pointer here
func (wo *WriterOptions) Apply(writer *storage.Writer, o *blob.WriterOptions) {
  writer.ChunkSize = wo.BufferSize
}
```

Reply:
Now I'm confused :)
It was mostly to differentiate between set/unset values similar to a lot of Kubernetes config structs (e.g. kubeadm) are doing it. I guess it's equivalent

Reply:
Would just mean it's not possible to set the values to the zero values on purpose but we probably don't need it as setting it to the zero value is equivalent to not setting the option at all.

Reply:
Should I just change the pointer fields to non-pointers and check for non-zero values instead of non-nil values?
If yes, should I change the receiver of Apply to a pointer receiver? (if yes, why)


I can also just overwrite the values in writer and blobl.WriterOptions independent of if they are zero values or not and change the parameter of the Writer method to a single WriterOptions struct and not a vararg of WriterOptions. There are currently no cases anymore where we hand over multiple WriterOptions (the With functions are unused). But this would in turn mean we have to handover nil everywhere to the Writer method because it's not a vararg anymore.

Reply:
This is different from the pattern used by storage:

https://github.com/googleapis/google-api-go-client/blob/v0.21.0/option/option.go#L17-L45

which would be something more along the lines of:

```golang

type WriterOption interface {
  Apply(writer *storage.Writer, o *blob.WriterOptions) error // I'd recommend allowing returning an error -- aka applying an AWS options to GCS
}

type contentEncoding struct {
  encoding string
}

func (ce contentEncoding) Apply(w *storage.Writer, o *blob.WriterOptions) error {
  if writer != nil { w.ContentEncoding = ce.encoding }
  return nil
}

func WithContentEncoding(encoding string) WriterOption {
  return contentEncoding{encoding}
}

// etc
```

Usage would then look like:

```golang
writer, err := opener.Writer(ctx, path, WithContentEncoding("text/plain"), WithBufferSize(1024*1024))
```

Thoughts on following this pattern?


IMO if you make the Writer signature:
```golang
func (o *opener) Writer(ctx context.Context, p string, opts ...WriterOptions) (io.WriteCloser, error) {
```

Then this decorator pattern makes more sense to me. If you want to do the `if opt.BufferSize != nil`  check, then the `...WriterOptions` seems unnecessary (since you can express everything in a single writeroptions)

Reply:
@fejta he did that because of my comment here: https://github.com/kubernetes/test-infra/pull/16722#discussion_r403507366

Reply:
This boils down to the fact that we have two opts structs which we need to manipulate, so we dont have a single place that contains all possible options anymore. Thats why a struct that has all possible options that are supported on both implementations makes IMHO sense.

The variadic arg makes sense even if you at most pass in one option, because it makes that particular argument optional.

Reply:
I would like to keep the single struct. @alvaroaleman pointed it out here: https://github.com/kubernetes/test-infra/pull/16722#discussion_r403507366 
I think it's better to have all options in one place. 

Okay so next try :). I replaced the vararg and the pointers in the options struct. Should make more sense now ;)

Reply:
Alvaro is right that to `Writer(ctx, p string, opts *WriterOptions)` breaks other users of the opener in a way that `...WriterOptions` does not, an makes it non-optional. The `...WriterOptions` seems superior to me. It makes the options actually optional

Can we put the `...` back?

Reply:
Ah @alvaroaleman didn't saw you're comment. Okay let's find consensus first. I can revert it back again or change it to some compromise in between

Reply:
btw. the variant with a single struct without pointers would be equivalent of the corresponding signature of gocloud (https://gocloud.dev/howto/blob/#writing)

Reply:
I think we should decide between:
* single struct with pointers for nil checking which is handed over via varargs
* single struct without pointers which is handed over as a pointer (we would just overwrite all values which are given without checking for nil or zero values)

Reply:
I'm fine with using a single struct, but I don't get why it is superior.

```golang
options := []io.WriterOption{
  io.WithContentType("text/plain"), 
  io.WithBufferSize(1024*1024),
}

// func WriterOptionsFromFileName(path string) []io.WriterOption
options = append(options, WriterOptionsFromFileName(path)...)

return opener.Writer(ctx, path, options...)
```

seems at least as clean to me as:

```golang

options := io.WriterOptions{
  ContentType: "text/plain",
  BufferSize: 1024*1024,
}

moreOptions := WriterOptionsFromFileName(path)

return opener.Writer(ctx, path, options, moreOptions) // how do we merge these two?
```

Hypothetically (not relevant now) it makes it more obvious which options are common `io.WithFoo()` and which are provider specific `s3.WithBar()`, `google.WithFrobber()`

Reply:
Probably a minor difference in that regard yup. Single struct is in exactly the same place in the options.go. The With* functions are distributed across the file. The advantage with s3.WithBar() only works if we create new packages for options of different providers, right?

Reply:
@fejta Fine for me change back to vararg. This also means reverting back to the pointers, right?

Reply:
```go
// how do we merge these two?
```
If we use vararg in both cases, the problem is the same right (if we're only applying values which are set (i.e. != nil)? If we pass in the same option multiple times the naive solution would be last one wins.

Reply:
Yes, I would consider `io.WithS3Bar(), io.WithGCSFoo()` and `WriterOptions{S3Bar: bar, GCSFoo: foo}` equivalent.

And yes, vararg and pointers or vararg and WithFoo() then how to merge makes sense in both examples.

Reply:
I reverted to a single struct with pointers handed over via varargs.